sound/soc/sof/sof-audio.h

```diff
@@ -465,6 +465,54 @@ struct snd_sof_route {
 	void *private;
 };
 
+/**
+ * struct snd_sof_dai_dma_ops - DAI DMA ops optional by default unless specified otherwise
+ * @stream_get_dma: Function pointer to get DAI DMA data
+ * @stream_put_dma: Function pointer to put DAI DMA data
+ * @setup: Function pointer for DAI DMA setup
+ * @free: Function pointer for DAI DMA free
+ * @pre_trigger: Function pointer for DAI DMA pre-trigger actions
+ * @trigger: Function pointer for DAI DMA trigger actions
+ * @post_trigger: Function pointer for DAI DMA post-trigger actions
+ */
+struct snd_sof_dai_dma_ops {
+	void *(*stream_get_dma)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+				struct snd_pcm_substream *substream);
+	void (*stream_put_dma)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+			       struct snd_pcm_substream *substream);
+	int (*setup)(struct snd_sof_dev *sdev, void *dma_data, unsigned int val);
+	void (*free)(struct snd_sof_dev *sdev, void *dma_data);
+	int (*pre_trigger)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+			   struct snd_pcm_substream *substream, int cmd);
+	int (*trigger)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+		       struct snd_pcm_substream *substream, int cmd);
+	int (*post_trigger)(struct snd_sof_dev *sdev, struct snd_soc_dai *cpu_dai,
+			    struct snd_pcm_substream *substream, int cmd);
+};
+
+/**
+ * struct snd_sof_dai_codec_ops - Codec DAI ops optional by default unless specified otherwise
+ * @params: Function pointer for codec DAI params
+ * @free: Function pointer for codec DAI free
+ */
+struct snd_sof_dai_codec_ops {
+	int (*params)(struct snd_sof_dev *sdev, struct snd_soc_dai *codec_dai,
+		      void *dma_data, struct snd_pcm_substream *substream,
+		      struct snd_pcm_hw_params *params, unsigned int *val);
+	int (*free)(struct snd_sof_dev *sdev, struct snd_soc_dai *codec_dai,
+		    void *dma_data, struct snd_pcm_substream *substream);
+};
+
+/**
+ * struct snd_sof_dai_ops - DAI ops
+ * @dma_ops: Pointer to the DAI DMA ops
+ * @codec_ops: Pointer to the CODEC DAI ops
```

Comment:
oh this is so confusing.

Is this a DAI ops for codec DAIs only?
If yes, why do we care about DMA for codec dai?
If not, why do we care about the codec DAI ops?

Lost.